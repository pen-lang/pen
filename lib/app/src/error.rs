use std::{error::Error, fmt::Display};

#[allow(clippy::enum_variant_names)]
#[derive(Debug)]
pub enum ApplicationError {
    MainFunctionTypeNotFound,
    ModuleNotFound(String),
    PackageNotFound(String),
    SystemPackageNotFound,
}

impl Error for ApplicationError {}

impl Display for ApplicationError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match self {
            Self::MainFunctionTypeNotFound => {
                write!(formatter, "main function type not found")
            }
            Self::ModuleNotFound(module) => {
                write!(formatter, "module {} not found", module)
            }
            Self::PackageNotFound(package) => {
                write!(formatter, "package {} not found", package)
            }
            Self::SystemPackageNotFound => {
                write!(formatter, "system package not found")
            }
        }
    }
}
