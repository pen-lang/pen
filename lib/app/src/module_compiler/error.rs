use std::{error::Error, fmt::Display};

#[derive(Debug)]
pub enum ModuleCompilerError {
    MainFunctionTypeNotFound,
}

impl Error for ModuleCompilerError {}

impl Display for ModuleCompilerError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match self {
            Self::MainFunctionTypeNotFound => {
                write!(formatter, "main function type not found")
            }
        }
    }
}
