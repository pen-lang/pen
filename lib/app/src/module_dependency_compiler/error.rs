use std::{error::Error, fmt::Display};

#[derive(Debug)]
pub enum ModuleDependencyCompilerError {
    PackageNotFound(String),
}

impl Error for ModuleDependencyCompilerError {}

impl Display for ModuleDependencyCompilerError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match self {
            Self::PackageNotFound(package) => {
                write!(formatter, "package {} not found", package)
            }
        }
    }
}
