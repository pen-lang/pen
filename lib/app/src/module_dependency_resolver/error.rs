use std::{error::Error, fmt::Display};

#[derive(Debug)]
pub enum ModuleDependencyResolverError {
    PackageNotFound(String),
}

impl Error for ModuleDependencyResolverError {}

impl Display for ModuleDependencyResolverError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match self {
            Self::PackageNotFound(package) => {
                write!(formatter, "package {} not found", package)
            }
        }
    }
}
