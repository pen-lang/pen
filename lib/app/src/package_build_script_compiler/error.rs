use std::{error::Error, fmt::Display};

#[derive(Debug)]
pub enum PackageBuildScriptCompilerError {
    SystemPackageNotFound,
}

impl Error for PackageBuildScriptCompilerError {}

impl Display for PackageBuildScriptCompilerError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match self {
            Self::SystemPackageNotFound => {
                write!(formatter, "system package not found")
            }
        }
    }
}
