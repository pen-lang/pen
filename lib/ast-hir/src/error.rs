use position::Position;
use std::{
    error::Error,
    fmt::{self, Display, Formatter},
    num::{ParseFloatError, ParseIntError},
};

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum CompileError {
    ModuleNotFound(ast::ModulePath),
    ParseFloat {
        error: ParseFloatError,
        position: Position,
    },
    ParseInteger {
        error: ParseIntError,
        position: Position,
    },
    TooFewBranchesInIf(Position),
}

impl Display for CompileError {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        match self {
            Self::ModuleNotFound(path) => {
                write!(formatter, "module \"{}\" not found", path)
            }
            Self::ParseFloat { error, position } => {
                write!(formatter, "{}\n{}", error, position)
            }
            Self::ParseInteger { error, position } => {
                write!(formatter, "{}\n{}", error, position)
            }
            Self::TooFewBranchesInIf(position) => {
                write!(
                    formatter,
                    "too small number of branches in if\n{}",
                    position
                )
            }
        }
    }
}

impl Error for CompileError {}
