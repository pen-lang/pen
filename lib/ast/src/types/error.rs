use position::Position;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
pub struct Error {
    position: Position,
}

impl Error {
    pub fn new(position: Position) -> Self {
        Self { position }
    }

    pub fn position(&self) -> &Position {
        &self.position
    }

    pub fn set_position(mut self, position: Position) -> Self {
        self.position = position;
        self
    }
}
