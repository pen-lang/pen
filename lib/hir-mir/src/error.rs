use hir::{analysis::types::TypeError, ir::*, types};
use position::Position;
use std::{
    error::Error,
    fmt::{self, Display, Formatter},
};

#[derive(Clone, Debug, PartialEq)]
pub enum CompileError {
    AnyEqualOperation(Position),
    AnyTypeBranch(Position),
    CompileConfigurationNotProvided,
    DuplicateFunctionNames(Position, Position),
    DuplicateTypeNames(Position, Position),
    FunctionEqualOperation(Position),
    FunctionExpected(Position),
    InvalidRecordEqualOperation(Position),
    InvalidTryOperation(Position),
    ListExpected(Position),
    MainFunctionNotFound(Position),
    MirTypeCheck(mir::analysis::TypeCheckError),
    MissingElseBlock(Position),
    NewContextFunctionNotFound(Position),
    RecordFieldPrivate(Position),
    RecordFieldUnknown(Position),
    RecordFieldMissing(Position),
    RecordExpected(Position),
    RecordNotFound(types::Record),
    SpawnOperationArguments(Position),
    TryOperationInList(Position),
    TypeAnalysis(TypeError),
    TypeNotFound(types::Reference),
    TypeNotInferred(Position),
    TypesNotComparable(Position),
    TypesNotMatched(Position, Position),
    UnionOrAnyTypeExpected(Position),
    UnionTypeExpected(Position),
    UnreachableCode(Position),
    VariableNotFound(Variable),
    WrongArgumentCount(Position),
}

impl Display for CompileError {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        match self {
            Self::AnyEqualOperation(position) => {
                write!(
                    formatter,
                    "equal operator cannot be used with any type\n{}",
                    position
                )
            }
            Self::AnyTypeBranch(position) => {
                write!(
                    formatter,
                    "any type cannot be used for downcast\n{}",
                    position
                )
            }
            Self::CompileConfigurationNotProvided => {
                write!(formatter, "compile configuration not provided")
            }
            Self::DuplicateFunctionNames(one, other) => {
                write!(formatter, "duplicate function names\n{}\n{}", one, other)
            }
            Self::DuplicateTypeNames(one, other) => {
                write!(formatter, "duplicate type names\n{}\n{}", one, other)
            }
            Self::FunctionEqualOperation(position) => {
                write!(
                    formatter,
                    "equal operator cannot be used with function type\n{}",
                    position
                )
            }
            Self::FunctionExpected(position) => {
                write!(formatter, "function expected\n{}", position)
            }
            Self::InvalidRecordEqualOperation(position) => {
                write!(
                    formatter,
                    "equal operator cannot be used with record type containing any or function types\n{}",
                    position
                )
            }
            Self::InvalidTryOperation(position) => {
                write!(
                    formatter,
                    "try operation cannot be used in function not returning error\n{}",
                    position
                )
            }
            Self::ListExpected(position) => {
                write!(formatter, "list expected\n{}", position)
            }
            Self::MainFunctionNotFound(position) => {
                write!(formatter, "main function not found\n{}", position)
            }
            Self::MirTypeCheck(error) => {
                write!(formatter, "failed to check types in MIR: {}", error)
            }
            Self::MissingElseBlock(position) => {
                write!(
                    formatter,
                    "missing else block in if-type expression\n{}",
                    position
                )
            }
            Self::NewContextFunctionNotFound(position) => {
                write!(formatter, "new context function not found\n{}", position)
            }
            Self::RecordFieldPrivate(position) => {
                write!(formatter, "private record field\n{}", position)
            }
            Self::RecordFieldUnknown(position) => {
                write!(formatter, "unknown record field\n{}", position)
            }
            Self::RecordFieldMissing(position) => {
                write!(formatter, "missing record field\n{}", position)
            }
            Self::RecordExpected(position) => {
                write!(formatter, "record expected\n{}", position)
            }
            Self::RecordNotFound(record) => write!(
                formatter,
                "record type \"{}\" not found\n{}",
                record.name(),
                record.position()
            ),
            Self::SpawnOperationArguments(position) => {
                write!(
                    formatter,
                    "lambda expression in spawn operation cannot have any argument\n{}",
                    position
                )
            }
            Self::TryOperationInList(position) => {
                write!(
                    formatter,
                    "try operation not allowed in list literal\n{}",
                    position
                )
            }
            Self::TypeAnalysis(error) => write!(formatter, "{}", error),
            Self::TypeNotFound(reference) => write!(
                formatter,
                "type \"{}\" not found\n{}",
                reference.name(),
                reference.position()
            ),
            Self::TypeNotInferred(position) => {
                write!(formatter, "type not inferred\n{}", position)
            }
            Self::TypesNotComparable(position) => {
                write!(formatter, "types not comparable\n{}", position)
            }
            Self::TypesNotMatched(lhs_position, rhs_position) => write!(
                formatter,
                "types not matched\n{}\n{}",
                lhs_position, rhs_position
            ),
            Self::UnionOrAnyTypeExpected(position) => {
                write!(formatter, "union or any type expected\n{}", position)
            }
            Self::UnionTypeExpected(position) => {
                write!(formatter, "union type expected\n{}", position)
            }
            Self::UnreachableCode(position) => {
                write!(formatter, "unreachable code\n{}", position)
            }
            Self::VariableNotFound(variable) => write!(
                formatter,
                "variable \"{}\" not found\n{}",
                variable.name(),
                variable.position()
            ),
            Self::WrongArgumentCount(position) => {
                write!(
                    formatter,
                    "wrong number of arguments in function call\n{}",
                    position
                )
            }
        }
    }
}

impl Error for CompileError {}

impl From<mir::analysis::TypeCheckError> for CompileError {
    fn from(error: mir::analysis::TypeCheckError) -> Self {
        Self::MirTypeCheck(error)
    }
}

impl From<TypeError> for CompileError {
    fn from(error: TypeError) -> Self {
        Self::TypeAnalysis(error)
    }
}
