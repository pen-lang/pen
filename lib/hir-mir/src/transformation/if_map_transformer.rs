use super::super::error::CompileError;
use crate::{context::CompileContext, downcast_compiler};
use hir::{
    analysis::AnalysisError,
    ir::*,
    types::{self, Type},
};

pub fn transform(context: &CompileContext, if_: &IfMap) -> Result<Expression, CompileError> {
    let configuration = &context.configuration()?.map_type;
    let position = if_.position();

    let key_type = if_
        .key_type()
        .ok_or_else(|| AnalysisError::TypeNotInferred(position.clone()))?;
    let value_type = if_
        .value_type()
        .ok_or_else(|| AnalysisError::TypeNotInferred(position.clone()))?;
    let any_type = Type::from(types::Any::new(position.clone()));

    Ok(IfType::new(
        if_.name(),
        Call::new(
            Some(
                types::Function::new(
                    vec![
                        types::Reference::new(&configuration.map_type_name, position.clone())
                            .into(),
                        any_type.clone(),
                    ],
                    any_type.clone(),
                    position.clone(),
                )
                .into(),
            ),
            Variable::new(&configuration.get_function_name, position.clone()),
            vec![
                if_.map().clone(),
                TypeCoercion::new(
                    key_type.clone(),
                    any_type.clone(),
                    if_.key().clone(),
                    position.clone(),
                )
                .into(),
            ],
            position.clone(),
        ),
        vec![IfTypeBranch::new(
            types::Reference::new(&configuration.empty_type_name, position.clone()),
            if_.else_().clone(),
        )],
        Some(ElseBranch::new(
            Some(types::Any::new(position.clone()).into()),
            Let::new(
                Some(if_.name().into()),
                Some(value_type.clone()),
                downcast_compiler::compile(
                    &any_type,
                    value_type,
                    &Variable::new(if_.name(), position.clone()).into(),
                    context,
                )?,
                if_.then().clone(),
                position.clone(),
            ),
            position.clone(),
        )),
        position.clone(),
    )
    .into())
}

#[cfg(test)]
mod tests {
    use super::*;
    use position::{test::PositionFake, Position};

    #[test]
    fn transform_if_map() {
        insta::assert_debug_snapshot!(transform(
            &CompileContext::dummy(Default::default(), Default::default()),
            &IfMap::new(
                Some(types::Number::new(Position::fake()).into()),
                Some(types::None::new(Position::fake()).into()),
                "x",
                Variable::new("xs", Position::fake()),
                Variable::new("k", Position::fake()),
                Variable::new("x", Position::fake()),
                None::new(Position::fake()),
                Position::fake(),
            ),
        ));
    }
}
