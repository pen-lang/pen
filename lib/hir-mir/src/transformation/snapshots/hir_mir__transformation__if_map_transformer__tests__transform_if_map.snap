---
source: lib/hir-mir/src/transformation/if_map_transformer.rs
assertion_line: 81
expression: "transform(&CompileContext::dummy(Default::default(), Default::default()),\n          &IfMap::new(Some(types::Number::new(Position::fake()).into()),\n                      Some(types::None::new(Position::fake()).into()), \"x\",\n                      Variable::new(\"xs\", Position::fake()),\n                      Variable::new(\"k\", Position::fake()),\n                      Variable::new(\"x\", Position::fake()),\n                      None::new(Position::fake()), Position::fake()))"
---
Ok(
    IfType(
        IfType {
            name: "x",
            argument: Call(
                Call {
                    function_type: Some(
                        Function(
                            Function {
                                arguments: [
                                    Reference(
                                        Reference {
                                            name: "GenericMap",
                                            position: Position {
                                                path: "",
                                                line_number: 0,
                                                column_number: 0,
                                                line: "",
                                            },
                                        },
                                    ),
                                    Any(
                                        Any {
                                            position: Position {
                                                path: "",
                                                line_number: 0,
                                                column_number: 0,
                                                line: "",
                                            },
                                        },
                                    ),
                                ],
                                result: Any(
                                    Any {
                                        position: Position {
                                            path: "",
                                            line_number: 0,
                                            column_number: 0,
                                            line: "",
                                        },
                                    },
                                ),
                                position: Position {
                                    path: "",
                                    line_number: 0,
                                    column_number: 0,
                                    line: "",
                                },
                            },
                        ),
                    ),
                    function: Variable(
                        Variable {
                            name: "getMap",
                            position: Position {
                                path: "",
                                line_number: 0,
                                column_number: 0,
                                line: "",
                            },
                        },
                    ),
                    arguments: [
                        Variable(
                            Variable {
                                name: "xs",
                                position: Position {
                                    path: "",
                                    line_number: 0,
                                    column_number: 0,
                                    line: "",
                                },
                            },
                        ),
                        TypeCoercion(
                            TypeCoercion {
                                from: Number(
                                    Number {
                                        position: Position {
                                            path: "",
                                            line_number: 0,
                                            column_number: 0,
                                            line: "",
                                        },
                                    },
                                ),
                                to: Any(
                                    Any {
                                        position: Position {
                                            path: "",
                                            line_number: 0,
                                            column_number: 0,
                                            line: "",
                                        },
                                    },
                                ),
                                argument: Variable(
                                    Variable {
                                        name: "k",
                                        position: Position {
                                            path: "",
                                            line_number: 0,
                                            column_number: 0,
                                            line: "",
                                        },
                                    },
                                ),
                                position: Position {
                                    path: "",
                                    line_number: 0,
                                    column_number: 0,
                                    line: "",
                                },
                            },
                        ),
                    ],
                    position: Position {
                        path: "",
                        line_number: 0,
                        column_number: 0,
                        line: "",
                    },
                },
            ),
            branches: [
                IfTypeBranch {
                    type_: Reference(
                        Reference {
                            name: "notFound",
                            position: Position {
                                path: "",
                                line_number: 0,
                                column_number: 0,
                                line: "",
                            },
                        },
                    ),
                    expression: None(
                        None {
                            position: Position {
                                path: "",
                                line_number: 0,
                                column_number: 0,
                                line: "",
                            },
                        },
                    ),
                },
            ],
            else_: Some(
                ElseBranch {
                    type_: Some(
                        Any(
                            Any {
                                position: Position {
                                    path: "",
                                    line_number: 0,
                                    column_number: 0,
                                    line: "",
                                },
                            },
                        ),
                    ),
                    expression: Let(
                        Let {
                            name: Some(
                                "x",
                            ),
                            type_: Some(
                                None(
                                    None {
                                        position: Position {
                                            path: "",
                                            line_number: 0,
                                            column_number: 0,
                                            line: "",
                                        },
                                    },
                                ),
                            ),
                            bound_expression: IfType(
                                IfType {
                                    name: "$value",
                                    argument: Variable(
                                        Variable {
                                            name: "x",
                                            position: Position {
                                                path: "",
                                                line_number: 0,
                                                column_number: 0,
                                                line: "",
                                            },
                                        },
                                    ),
                                    branches: [
                                        IfTypeBranch {
                                            type_: None(
                                                None {
                                                    position: Position {
                                                        path: "",
                                                        line_number: 0,
                                                        column_number: 0,
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            expression: Variable(
                                                Variable {
                                                    name: "$value",
                                                    position: Position {
                                                        path: "",
                                                        line_number: 0,
                                                        column_number: 0,
                                                        line: "",
                                                    },
                                                },
                                            ),
                                        },
                                    ],
                                    else_: None,
                                    position: Position {
                                        path: "",
                                        line_number: 0,
                                        column_number: 0,
                                        line: "",
                                    },
                                },
                            ),
                            expression: Variable(
                                Variable {
                                    name: "x",
                                    position: Position {
                                        path: "",
                                        line_number: 0,
                                        column_number: 0,
                                        line: "",
                                    },
                                },
                            ),
                            position: Position {
                                path: "",
                                line_number: 0,
                                column_number: 0,
                                line: "",
                            },
                        },
                    ),
                    position: Position {
                        path: "",
                        line_number: 0,
                        column_number: 0,
                        line: "",
                    },
                },
            ),
            position: Position {
                path: "",
                line_number: 0,
                column_number: 0,
                line: "",
            },
        },
    ),
)
